contracts/BatchAuction.sol

```diff
@@ -59,24 +65,45 @@ contract BatchAuction is Ownable {
 
         address tokenAddress = tokenIdToAddressMap[tokenIndex];
         require(tokenAddress != address(0), "Requested token is not registered");
-
         require(
             ERC20(tokenAddress).transferFrom(msg.sender, address(this), amount), 
             "Unsuccessful transfer"
         );
-        
-        if (depositIndex != block.number / 20) {
-            depositIndex = block.number / 20;
-            emit IncrementDepositIndex(depositIndex);
-        }
 
+        uint depositSlot = this.depositIndex();
         uint16 accountId = publicKeyToAccountMap[msg.sender];
         bytes32 nextDepositHash = sha256(
-            abi.encodePacked(depositHashes[depositIndex].shaHash, accountId, tokenIndex, amount)
+            abi.encodePacked(depositHashes[depositSlot].shaHash, accountId, tokenIndex, amount)
         );
 
-        depositHashes[depositIndex] = DepositState({shaHash: nextDepositHash, applied: false});
+        depositHashes[depositSlot] = DepositState({shaHash: nextDepositHash, applied: false});
+        emit Deposit(accountId, tokenIndex, amount, depositSlot);
+    }
+
+    function depositIndex() public view returns (uint) {
+        return block.number / 20;
+    }
+
+    function stateIndex() public view returns (uint) {
+        return stateRoots.length - 1;
+    }
 
-        emit Deposit(accountId, tokenIndex, amount, depositIndex);
+    function applyDeposits(
+        uint slot,
+        bytes32 _currDepositHash,
+        bytes32 _currStateRoot,
+        bytes32 _newStateRoot
+    )
+        public onlyOwner()
+    {   
+        require(slot < this.depositIndex(), "Deposit slot must exist and be inactive");
```

Comment:
@fleupold 

In response to your request

```
Can you add a unit test that makes sure we cannot do the following (in the same block)

Deposit a
ApplyDeposit a
Deposit b
Apply deposit b fails (since it is in the same deposit index as a and a has already been applied)
```
This constraint implies that there is no possible way to apply deposit A and have any other future deposits the same deposit index afterwards. 

Essentially, deposits can't be applied until the deposit index has become "inactive"

Reply:
In fact, there is already a unit test demonstrating that this scenario can not happen

https://github.com/gnosis/dex-contracts/blob/0dc02b0daf19b45b2eb6eccf5cf1af1da1686b52/test/auction_accounts.js#L221-L226

Reply:
Thanks for explaining why this is not possible. I didn't interpret this require correctly.
The existing test would fail regardless of this constraint though, since we are trying to applyDeposits on an empty block (which isn't possible in the first place).
Thus, we would lose line coverage if the require statements were reversed. Could you add a deposit call to the test to make it a bit more robust?

Reply:
Yes absolutely. I can used some of my non-trivial deposit test hashes from below! Good catch. Its sad that we can't somehow catch the revert message and assert we get the correct ones.